src/report_reader.cpp

```diff
@@ -284,24 +285,64 @@ std::vector<NodeID> ReportReader<T>::Population::getNodeIds() const {
 }
 
 template <typename T>
-Selection::Values ReportReader<T>::Population::node_ids_from_selection(
-    const nonstd::optional<Selection>& selection) const {
-    Selection::Values node_ids;
-
-    if (!selection) {  // Take all nodes in this case
-        node_ids.reserve(nodes_pointers_.size());
-        std::transform(nodes_pointers_.begin(),
-                       nodes_pointers_.end(),
-                       std::back_inserter(node_ids),
-                       [](const std::pair<NodeID, Range>& node_pointer) {
-                           return node_pointer.first;
-                       });
-    } else if (selection->empty()) {
-        return {};
-    } else {
-        node_ids = selection->flatten();
+std::tuple<NodePointers, Range, typename DataFrame<T>::DataType>
+ReportReader<T>::Population::getElementIds(const nonstd::optional<Selection>& selection) const {
+    NodePointers node_pointers;
+    Range range = {std::numeric_limits<uint64_t>::max(), std::numeric_limits<uint64_t>::min()};
+    typename DataFrame<T>::DataType ids{};
+    size_t eids_count = 0;
+
+    // Helper function to update the min / max values and count
+    const auto& update_range = [&range, &eids_count](const Range& range_new) {
+        range.first = std::min(range.first, range_new.first);
+        range.second = std::max(range.second, range_new.second);
+        eids_count += (range_new.second - range_new.first);
+    };
+
+    // Take all nodes if no selection is provided
+    if (!selection) {
+        node_pointers = nodes_pointers_;
```

Comment:
Yes, yes it is. The original implementation splits the `NodeID` keys of the map into an `std::vector`, and then the positions into another vector. The current change unifies the data structures for simplicity.

We thought about using an `std::shared_ptr` of the `node_pointers_` map, allowing us to just point to the original when this is "all the GIDs" (constant time) or to a new map when it is not. However, the performance of the copy for the `std::map` here is almost negligible compared to the rest of the code and it would make the source code more difficult to read.

Reply:
This has changed now, closing the conversation.